[2단계 - 자동차 경주 리팩터링] 디우(김동규) 미션 제출합니다.

tco0427 opened the pull request:
## 안녕하세요 제이!! 디우입니다ㅎㅎ

- 2단계에서 요구사항은 MVC 패턴 기반으로 리팩토링하는 것이었는데요! 기존의 RacingGame이 Controller의 역할을 하고 있으면서 `view 패키지의 객체가 domain 패키지 객체에 의존할 수 있지만, domain 패키지의 객체는 view 패키지 객체에 의존하지 않도록 구현` 한다는 내용을 부합하고 있다고 판단하여 크게 구조가 바뀌지는 않았습니다!! 
- 또한 1단계 마지막으로 피드백 주신 부분을 반영해보았습니다! 보내주신 '일급 컬렉션'에 대한 내용의 글도 읽어보며 리팩토링 해보았습니다! 그러면서 가장 크게는 ValidationUtil 클래스를 제거한 것인데요! 예를 들어 Car를 생성할 때 검증이 이루어져야 하기 때문에 강하게 연결되어있다는 생각이 들었고, 해당 클래스 내부로 포함시켰습니다!! (InitUtil도 마찬가지의 이유로 도메인의 포함시켰습니다. 또한 전체 시도 횟수도 말씀해주신 것처럼 하나의 객체로 포장해보았습니다.ㅎㅎ 더욱 깔끔해진 것 같아요!)
- 이외에도 우승자가 한 명인 경우와 우승자가 복수인 경우에 대해서도 테스트할 수 있도록 단위 테스트 코드를 작성해보았습니다! 이를 통해서 자동차들에 대한 검증도 이루어질 수 있도록 해보았습니다. 제 생각에는 해당 애플리케이션에 대한 충분한 테스트가 되었다고 생각하지만...제가 생각지 못한 부분이 아직 있을 수 있다고 생각합니다..ㅜ.ㅜ 피드백 부탁드립니다!!

### 따끔한 피드백 부탁드립니다!! 감사합니다!

tco0427 committed:
fix: 접근제어자 변경

public 으로 선언된 접근제어자에 대해서 private 으로 통일되게 수정하였습니다.

tco0427 committed:
refactor: 불필요한 할당문 제거

carList 필드에서 new ArrayList<>() 를 통한 생성을 제거하여 중복된 생성을 제거하였습니다. 이를 통해 객체의 생성주기를 객체가 필요한 시점으로 설정할 수 있습니다.

tco0427 committed:
refactor: List의 sort() 이용하도록 수정

Collections에서 제공하는 sort() 대신 List 인터페이스에서 제공하는 sort() 이용

tco0427 committed:
style: formatting 통일

RandomUtil 클래스의 formatting을 다른 클래스(코드)와 통일해주었습니다.

tco0427 committed:
refactor: 불필요한 import문 제거

불필요한 import구문을 제거하였습니다.

tco0427 committed:
refactor: 일급 컬렉션 사용

일급 컬렉션을 사용하여 CarList를 분리하여 클래스화

tco0427 committed:
refactor: totalAttemptCount 클래스화

게임 전체 시도횟수에 해당하는 totalAttemptCount 를 클래스화하였습니다.

tco0427 committed:
fix: 반환형 수정

initTotalAttempt() 메소드 반환형을 int에서 void로 수정하였습니다.

tco0427 committed:
fix: 테스트코드 수정

GameTotalCount로 totalAttemptCount를 분리해냄에 따른 테스트 코드 수정

tco0427 committed:
refactor: 테스트 코드 분리

시도 횟수와 관련된 테스트를 GameTotalCountTest로 분리하였습니다.

tco0427 committed:
test: 우승자 테스트

우승자가 한 명인 경우와 복수인 경우에 대해서 테스트하였습니다.

tco0427 committed:
refactor: 패키지명 변경

RacingGame 이 Controller 의 역할을 하고 있다고 판단되었고, 이를 명확히 하기 위해서 package 명을 변경하였습니다.

tco0427 committed:
refactor: 자동차 이름 검증 Car 내부로 병합

Cars 클래스와 통일성 있게 VerificationUtil 에 있던 자동차 이름 검증 코드를 Car 클래스 내부로 병합하였습니다.

tco0427 committed:
refactor: Car 변경에 의한 CarTest 수정

Car 클래스와 VerificationUtil 삭제에 따른 Test 코드 수정을 진행하였습니다.

jjj0611 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -8,7 +9,7 @@
 class CarTest {
 
-    public static final int MOVE_CONDITION= 10;
+    public static final int MOVE_CONDITION = 10;
```
디우 여기서 왜 이동 조건과 멈춤 조건을 각각 0과 10으로 설정했을까요?
테스트는 하나의 좋은 문서가 되어야 합니다.
만약 이 테스트 코드를 읽은 사람은 
아 자동차는 10이라는 숫자를 받으면 이동하는구나!
아 자동차는 0이라는 숫자를 받으면 멈추는구나!
하고 이해하게 될거에요!

테스트가 좋은 문서가 되게 테스트를 수정해보면 어떨까요?

jjj0611 replied:
더불어서 public으로 설정한 이유가 있나요??

tco0427 replied:
```java
private static final int MINIMUM_MOVE_CONDITION = 4;
private static final int MAXIMUM_STOP_CONDITION = 3;
```
이 보다 적절한 네이밍이라고 생각되어 수정해보았습니다. 더불어 private으로 접근제어자 수정하였습니다. (cmd + option + c 로 생성시 public으로 되어있는 것 같네요...주의하겠습니다!!ㅜㅜ)

jjj0611 replied:
네이밍 뿐만 아니라 10과 0이라는 숫자 대신 3과 4라는 엣지 케이스에 대해서 테스트를 해달라고 요청 드린건데 잘 반영해주셨네요 :)
테스트를 할 땐 항상 엣지케이스로 테스트하는 것을 고민하시면 좋습니다 :)

jjj0611 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -52,4 +53,34 @@
+
+    @Test
+    @DisplayName("다섯 글자 이상의 자동차 이름 테스트")
+    public void 다섯글자_이상의_자동차_이름_테스트() {
```
다섯글자 '이상'인가요??

tco0427 replied:
다섯 글자 초과이네요..놓친 부분이었습니다..감사합니다!!😃

jjj0611 commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,65 @@
+class CarsTest {
+
+    public static final int MOVE_CONDITION = 10;
+    public static final int WINNER_INDEX = 0;
```
ditto

tco0427 replied:
private으로 수정하면서 함께 MINIMUM_MOVE_CONDITION으로 수정하였습니다!!

jjj0611 commented on src/test/java/racingcar/domain/GameTotalCountTest.java:
```diff
@@ -0,0 +1,29 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class GameTotalCountTest {
```
정상 케이스에 대한 것도 추가가 되어야 하지 않을까요?!

tco0427 replied:
정수로 변환이 가능한 문자열이 입력되었을 때에 대한 테스트도 추가해보았습니다!!

jjj0611 commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,65 @@
+
+        assertThat(winner.size()).isEqualTo(1);
+        assertThat(winner.get(0)).isEqualTo("dwoo");
+        assertThat(winnerPosition).isEqualTo(cars.getCar(WINNER_INDEX).getPosition());
```
테스트 케이스를 아주 잘 작성해주셨지만,
조금 더 읽기 좋은 방향으로 시도할 수 없을까요?

tco0427 replied:
단일 우승자와 복수 우승자에 대한 테스트에서 중복되는 부분을 private 메소드로 추출하고, 동일한 given절을 @BeforeEach 로 빼서 보다 가독성있게 수정해보았습니다. 이부분은 조금 더 고민해보도록 하겠습니다!

jjj0611 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -1,32 +1,29 @@
+
         while (totalAttemptCount --> ZERO) {
             run();
         }
```
객체에서 꺼내지 말고 메시지를 던져보면 어떨까요?
[이 글](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/)도 한 번 참고해보시면 좋을 것 같아요 :)

tco0427 replied:
아하! 좋은 방법인 것 같습니다!! 적용해보았습니다~_~

jjj0611 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -1,32 +1,29 @@
 
 public class RacingGame {
 
     private static final int ZERO = 0;
```
0이 ZERO인 것은 알 것 같은데 0이 어떤 의미일까요?!

tco0427 replied:
게임 종료 조건을 의미하는 0입니다..!! 의미를 나타낼 수 있도록 상수명을 수정해보았습니다!

jjj0611 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,11 +1,11 @@
-
 public class Car {
 
     private static final int INITIAL_VALUE = 0;
```
무엇의 '초기 값'인가요?!
의미를 좀 명확하게 가져가면 어떨까요?!

tco0427 replied:
마찬가지로 position에 대한 초기값이므로 의미를 보다 명확히 하기 위해 상수명을 변경하였습니다!! 피드백 감사합니다!

jjj0611 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,11 +1,11 @@
+    private static final int CAR_NAME_MAX_LENGTH = 5;
 
     private final String name;
     private int position = INITIAL_VALUE;
```
시도 횟수를 감싼 것처럼 감싸볼 수 있지 않을까요?!

tco0427 replied:
CarName이라는 새로운 클래스로 감싸보았습니다😄

jjj0611 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,11 +1,11 @@
 public class Car {
 
     private static final int INITIAL_VALUE = 0;
     private static final int PROGRESS_CONDITION_VALUE = 4;
```
자동차의 진행을 결정하는 로직을 인터페이스로 분리해서 구현해보면 어떨까요?

tco0427 replied:
해당 내용에 대해서 이해가 잘 가지 않아요ㅠㅠ...

jjj0611 replied:
이 부분은 DM으로 논의 했으니 넘어갈게요!

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,67 @@
+        initCar(carNames);
+    }
+
+    public int getMaxPosition() {
+        cars.sort((o1, o2) -> o2.getPosition() - o1.getPosition());
+
+        return cars.get(FIRST_INDEX).getPosition();
+    }
+
+    public List<String> getWinner(int position) {
+        return cars.stream()
+                .filter(car -> car.isSamePosition(position))
+                .map(car -> car.getName())
+                .collect(Collectors.toList());
+    }
```
두 메서드를 따로 분리하신 이유가 있을까요??

tco0427 replied:
현재 cars 리스트에서 최대 position값을 얻는 것과 우승자 이름 리스트를 얻는 것은 서로 다른 작업이라고 생각하였고, 이 두 로직이 하나의 메소드에 있는 것은 하나의 메소드에서 두가지 일을 하는 것이라고 생각하여 분리하였습니다..!!

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public int getSize() {
+        return cars.size();
+    }
+
+    public Car getCar(int index) {
+        return cars.get(index);
+    }
```
이렇게 됐을 때 List<Car>를 반환하는 것과 어떤 차이가 있을까요?

tco0427 replied:
음...데이터(car 리스트) 자체를 반환하여 외부에 드러내느냐 그렇지 않느냐는 차이가 있다고 생각합니다..!

jjj0611 commented on src/main/java/racingcar/domain/GameTotalCount.java:
```diff
@@ -0,0 +1,41 @@
+package racingcar.domain;
+
+public class GameTotalCount {
```
잘 감싸주셨지만 생성에 대한 로직 말고 게인 진행에 대한 로직도 함께 가져갈 수 있을 것 같아요!

tco0427 replied:
isContinue()라는 메소드로 게임 진행 여부에 대한 로직을 포함시켜 보았습니다!!

jjj0611 reviewed (changes requested):
안녕하세요 디우! 리뷰어 제이입니다.
간단하게 코멘트 드렸어요!
확인하시고 반영해주세요

더불어서 게임 횟수로 2147483648를 넣으면 실패하네요!
확인 부탁드려요 :)

tco0427 committed:
refactor: Test 코드의 상수에 대해서 private

Test 코드의 상수에 대해서 public으로 열어둘 필요가 없으므로 private으로 수정하였습니다.

tco0427 committed:
refactor: Test 코드 DisplayName 변경

요구사항이 5글자 이하까지 가능하므로 `5글자 이상`에 대한 테스트가 아닌 `5글자 초과`에 대한 테스트 이름이 더 적절하므로 수정하였습니다.

tco0427 committed:
refactor: 자동차 이동에 관한 상수명 수정

요구사항이 0~9까지의 랜덤값에서 4이상인 경우에 대해서 전진이므로 상수명을 MINIMUM_MOVE_CONDITION 과 MAXIMUM_STOP_CONDITION 으로 수정하여 테스트가 하나의 문서로써 역할을 할 수 있도록 수정하였습니다.

tco0427 committed:
test: 정상적인 시도횟수 입력 테스트

정상적인 시도횟수 입력이 제대로 동작하는지 테스트하는 테스트 코드를 추가하였습니다.

tco0427 committed:
refactor: 게임 진행을 GameTotalCount 에 메시지 보내도록 수정

게임 진행 여부에 대한 정보를 GameTotalCount 에서 데이터를 꺼내와서 검사하는 것이 아니라 메시지를 보내 검사하도록 수정해보았습니다.

tco0427 committed:
refactor: Car 클래스의 상수를 좀 더 명확하게 수정

Car 클래스의 INITIAL_VALUE 가 `무엇의` 초기값인지 의미가 불분명하므로 POSITION_INITIAL_VALUE 로 수정하였습니다.

tco0427 committed:
refactor: CarsTest 리팩터링

게임 진행을 하는 로직을 별도의 메소드로 분리하여 조금 더 가독성 있는 테스트 케이스가 될 수 있도록 수정해보았습니다.

tco0427 committed:
refactor: CarsTest 가독성 있게 리팩토링

@BeforeEach를 사용하여 보다 가독성 있는 단위 테스트가 되도록 리팩터링해보았습니다.

tco0427 committed:
fix: 누락된 @SuppressWarnings 추가

누락된 @SuppressWarnings("NonAsciiCharacters") 어노테이션을 추가해주었습니다.

tco0427 committed:
refactor: 자동차 이름을 별도의 클래스로 분리

자동차 이름 그리고 그에 대한 검증이 이루어지는 부분을 별도의 클래스로 분리하였습니다.

tco0427 commented:
피드백 주신 부분에 대해서 리팩터링을 진행해보았습니다!!
상수에 대한 private 접근제어자 설정이 계속 누락되네요.(cmd + option + c 로 생성시 public이 default라 제가 자꾸 놓치는것 같습니다ㅠㅠ 앞으로 꼼꼼히 확인하도록 하겠습니다...!!죄송합니다ㅠㅠ)
그런데 `자동차의 진행을 결정하는 로직을 인터페이스로 분리해서 구현해보면 어떨까요?`에 대한 코멘트가 이해가 잘 가지 않아서요..혹시 설명 한 번만 더 부탁드려도 될까요...??

tco0427 committed:
refactor: 자동차 이동 로직에 대해서 인터페이스로 분리해서 구현

자동차 이동 로직에 대해서 인터페이스로 분리해서 구현하였습니다. 이를 통해서 이동 로직의 변화에 대해서 Car는 영향을 받지 않도록 유연하게 구현해보았습니다.

tco0427 committed:
refactor: 전략 패턴 사용에 따른 테스트 코드 수정

전략 패턴을 사용함에 따른 단위 테스트 코드를 수정하였습니다.

tco0427 committed:
refactor: 불필요한 상수 제거

Car 클래스에서 사용하지 않는 상수 PROGRESS_CONDITION_VALUE 를 제거하였습니다.

tco0427 committed:
refactor: MovingStrategy 생성을 한번만 하도록 리팩터링

반복적인 new 키워드를 통한 MovingStrategy 구현체 생성보다는 한번의 생성이 보다 적절하다고 생각되어 수정하였습니다.

tco0427 committed:
refactor: 불필요한 import 구문 제거

Cars에서 더이상 RandomMovingStrategy에 대한 의존성을 가지지 않으므로 import 구문 제거

jjj0611 commented on src/main/java/racingcar/domain/GameTotalCount.java:
```diff
@@ -0,0 +1,52 @@
+        totalAttemptCount = totalAttemptCount -1;
+
+        return true;
+    }
```
현재 진행 중인 상태를 묻는 것과 상태를 변화시키는 메서드가 함께 존재하는 것 같아요. 이 부분이 가지고 올 수 있는 사이드 이펙트는 없을까요?

tco0427 replied:
전체 시도횟수에 대한 로직이 변경 -> 수정해야합니다!
진행 조건이 변경 -> 수정해야합니다!
변경의 원인이 여러개라고 생각합니다.
예를 들어서 전체 시도횟수에 대한 로직이 다음과 같이 바뀌면 isContinue를 변경해야합니다!
(VIP는 1씩 감소하고 GENERAL은 2씩 감소한다.)
별도의 메소드로 빼낼 수 있을 것 같습니다..!!

(제가 제대로 이해한 것인지 모르겠네요..ㅜㅜ)

jjj0611 replied:
아주 제대로 이해하셨어요 :)
진짜 진짜 너무 이른 코멘트일 수도 있지만,
명령과 조회의 분리라는 CQRS에 대해서 간단히만 찾아보세요!
너무 깊게 고민하진 마시고 아 이런 개념도 있구나 정도로 넘어가주시면 좋을 것 같아요 :)

jjj0611 commented on src/main/java/racingcar/domain/GameTotalCount.java:
```diff
@@ -0,0 +1,52 @@
+
+    public GameTotalCount(String attempt) {
+        initTotalAttempt(attempt);
+    }
```
디우 이 부분도 한 번 고민해보면 좋을 것 같아요.
GameTotalCount에서 가장 중요한 생성 규칙은 무엇일까요?
바로 0보다 큰 수를 기준으로 만드는 것입니다.
그렇다면 그것을 핵심 로직으로 두고, 문자열 등 다른 입력 값으로 만드는 것을 다르게 구현하면 어떻까요?
이해가 어렵다면 Integer.valueOf를 한 번 참고해보세요 :)

tco0427 replied:
정적 팩토리 메소드! 새롭게 알게 된 개념입니다. 적용해보았습니다!!
관련 블로그 글을 일급컬렉션 때 처럼 정리하여 확실하게 내 것으로 만들 수 있도록 하겠습니다!! 감사합니다~_~

jjj0611 commented on src/main/java/racingcar/domain/GameTotalCount.java:
```diff
@@ -0,0 +1,52 @@
+    private void initTotalAttempt(String attempt) throws IllegalArgumentException {
+        int validateAttempt = validateAttempt(attempt);
+        this.totalAttemptCount = validateAttempt;
+    }
```
이건 생성자 안에 있어야 더 가독이 좋을 것 같은데 어떻게 생각하시나요??

tco0427 replied:
동의합니다!! 조금 더 가독성 있게 수정해보았습니다!

jjj0611 commented on src/main/java/racingcar/domain/movestrategy/RandomMovingStrategy.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public boolean isMovable() {
+        int randomNumber = RandomUtil.getRandomNumber();
```
RandomStrategy를 만들었는데, 그럼 RandomUtil에 있던 로직을 가져오면 어떨까요??

tco0427 replied:
동의합니다! 굳이 RandomUtil에 대한 의존성을 가지지 않고, 가져와서 private 메소드로 RandomMovingStrategy 내부에서만 사용해도 될 것 같아요!!

jjj0611 commented on src/main/java/racingcar/util/RandomUtil.java:
```diff
@@ -2,12 +2,12 @@
 public class RandomUtil {
 
-    public static final int RANDOM_RANGE= 10;
+    public static final int RANDOM_RANGE = 10;
```
public 이네요 🥲

tco0427 replied:
최대한 꼼꼼히 본다고 해봤는데...또 이렇게 놓친부분이...확인해주셔서 감사합니다..!!

jjj0611 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -46,10 +44,40 @@
         //when
         for (int i = 0; i < repeatCount; i++) {
-            car.progress(MOVE_CONDITION);
+            car.progress(() -> true);
```
인터페이스로 변경하니까 훨씬 깔끔해졌네요!
더불어서 람다 표현식에도 익숙해지고 있는 것 같아서 좋습니다!
lambda는 꼭 익숙해지시면 좋을 것 같아요 :)

tco0427 replied:
해당 부분이 약한 부분이었는데..많이 공부하고 있습니다!! 감사합니다~_~

jjj0611 reviewed (changes requested):
안녕하세요 디우! 디엠으로 적용할 것들에 대해 이야기 했는데 매우 잘 적용해주셨네요 :)
점점 코드도 많이 깔끔해지고 있는 것 같아요!

조금 더 적용해보면 좋을만한 부분 코멘트 드렸으니 확인 부탁드려요 :)

tco0427 committed:
refactor: RandomUtil 로직을 RandomMovingStrategy 내부로

RnadomMovingStrategy가 굳이 랜덤 값을 생성해주는 부분에 대한 로직을 가지고 있던 RandomUtil에 대한 의존성을 가지지 않고, 내부로 가지고와서 내부에서만 사용해도 된다고 생각하여 리팩터링 하였습니다.

tco0427 committed:
refactor: GameTotalCount 생성자 수정

initTotalAttempt() 메소드 내용을 생성자로 가져왔습니다. 이 과정을 통해서 보다 가독성 있는 GameTotalCount가 될 수 있습니다.

tco0427 committed:
refactor: GameTotalCount 리팩터링

게임 진행 가능 여부를 묻는 것과 이를 통해서 상태를 변화시키는 로직이 하나의 메소드에 있던 것을 분리하여 이를 통해서 발생할 수 있는 사이드 이펙트에 대응하였습니다.

tco0427 committed:
refactor: GameTotalCount 메소드 리팩터링

GameTotalCount 클래스의 양의 정수를 검증해 얻는 로직에 대한 리팩터링을 진행하였습니다. 생성자에서 호출하는 validateAttempt(시도횟수 검증) 이라는 메소드명보다는 translatePositiveInteger(양의 정수로 변환)이 더욱 적절하다고 판단되었고, 이에 따라 메소드에 대한 리팩터링을 진행하였습니다.

tco0427 committed:
refactor: 정적 팩토리 메소드를 활용하여 GameTotalCount 수정

GameTotalCount 는 `숫자` 라는데에 본질이 있다. 따라서 해당 생성자에서 숫자를 받고 0보다 큰 수인지를 체크하도록 하고 정적 팩토리 메소드를 두어 객체를 생성할 때 문자열에 대한 변환을 하도록 하여 분리하였습니다.
네이밍은 `정적 팩토리 메소드 네이밍 컨벤션`에 따라 `create`를 사용하였습니다.
(추가적으로 이에 따른 변경되는 변경점들을 함께 수정하였습니다.)

tco0427 committed:
refactor: translateInteger 로 네이밍 변경

보다 적절한 네이밍으로 수정하였습니다.

jjj0611 commented on src/main/java/racingcar/domain/GameTotalCount.java:
```diff
@@ -0,0 +1,51 @@
+
+        reduceAttemptCount();
+
+        return true;
```
엇 이해는 제대로 하신 것 같은데 반영이 안된 것 같아요 
맞을까요?

tco0427 replied:
다시 한 번 수정해보았습니다!

jjj0611 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,64 @@
+
+public class RacingGame {
+
+    private Cars cars;
+    private GameTotalCount gameTotalCount;
+    private MovingStrategy movingStrategy = new RandomMovingStrategy();
```
이 친구들을 멤버 변수로 놓지 않고 프로그래밍 해볼까요 디우!
그리고 한 번 뭐가 바뀌는지를 봐보면 좋을 것 같아요 :)

tco0427 replied:
저는 지역변수로 바꾸는 것을 생각해보았는데요! 여러가지가 바뀌는 것 같습니다..!!

우선 첫번째로 저의 경우 Cars에 대한 생성, GameTotalCount에 대한 생성 로직이 분리되었습니다.
다음으로 run, printProgress 등의 메소드에서 계속해서 매개변수로 해당 데이터를 넘겨주어야하다보니 코드가 지저분해졌습니다..
따라서 저는 해당 참조변수들은 멤버변수로 두는 것이 적절하다고 생각합니다..!!
해당 클래스의 멤버 변수(데이터)와 메소드(해당 데이터를 조작하는 코드)가 하나의 클래스에 모여있는 점에서도 객체 지향을 잘 따르게 있다고 생각되어..이를 어떤 지역변수로 바꿀 필요는 없다는 생각이 들었습니다...

jjj0611 replied:
넵 요 부분은 레벨2로 가면서 좀 더 깊게 고민해보면 좋을 것 같아요 :)

jjj0611 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,64 @@
+            int position = car.getPosition();
+
+            OutputView.printProgress(carName, position);
+        }
```
만약 자동차에 대한 데이터가 더 많아지면, 어떻게 될까요?
String carName = car.getName();
int position = car.getPosition();
String teamName = car.getTeamName();
...

OutputView.printProgress(carName, position, teamName, ...);

등 엄청난 증식체가 나타날겁니다... 😨

어떻게하면 좋을까요? :)

tco0427 replied:
Car를 넘겨주는 방법을 생각해보았습니다. 하지만 만약 domain에 해당하는 Car 객체를 그대로 view에 전달한다면 향후에 view에서는 불필요한 데이터(이 Car를 DB에 저장한다고 하면 id값 혹은 다른 entity와의 연관관계를 위한 데이터 등)를 함께 전달해야하므로 Car로부터 view에 필요한 데이터만을 가지는 CarDto를 만들어 넘겨주도록 하였습니다.

jjj0611 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,30 +1,28 @@
 
-    public void progress(int number) {
-        if (number >= PROGRESS_CONDITION_VALUE) {
+    public void progress(MovingStrategy movingStrategy) {
+        if (movingStrategy.isMovable()) {
```
자 이제 4라는 기준 값도 Car로부터 빠져나왔고,
이 기준이 4에서 5로 변한다고 하더라도 모든 테스트 코드들이 변하지 않아도 될거에요

기존에는 모든 테스트에서 car.progress(4);를 해서 테스트를 하고 있었기 때문에
이 기준값이 5로 변한다고 하면 모든 테스트가 깨지는 현상이 있었을 겁니다 :)

깔끔하고 좋은 코드의 특징 중 하나는 변경에 대한 사이드 이펙트가 최소화 된다는 점이라고 생각해요!
항상 변경에 유연한 코드를 고민하며 프로그래밍 해보면 좋을 것 같아요 :)

tco0427 replied:
아직은 여러 상황들을 스스로 생각하며 변경에 대해 대응하는 것이 솔직히 어렵기는 하지만...스스로 고민해보는 시간을 많이 가져보면서 성장해보도록 하겠습니다!! 감사합니다..!!

jjj0611 commented on src/main/java/racingcar/domain/CarName.java:
```diff
@@ -0,0 +1,43 @@
+    private static final int CAR_NAME_MIN_LENGTH = 1;
+    private static final int CAR_NAME_MAX_LENGTH = 5;
+
+    private String name;
```
변결될 가능성이 있을까요??

tco0427 replied:
자동차 이름의 길이에 대한 조건이 변경되어도 CarName을 사용하는 코드들에서는 영향을 받지 않고 코드가 변경될 필요가 없다고 생각합니다!
(CarName에서는 당연히 조건에 해당되는 상수와 ERROR 메세지에는 변경이 필요하지만 핵심 로직에는 영향이 없다고 생각합니다!)

jjj0611 replied:
앗! 이 부분은 name에 대한 final 키워드를 사용하면 어떨까하는 마음에 달았던 코멘트였어요!
제가 너무 추상적으로 코멘트를 드렸네요 😅

tco0427 replied:
아하! 이해하였습니다!

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,67 @@
+    private static final String SEPARATOR = ",";
+    private static final int FIRST_INDEX = 0;
+
+    private List<Car> cars;
```
변경될 가능성이 있을까요?

tco0427 replied:
아래 코멘트처럼 말씀해주신대로 SEPARATOR를 view쪽으로 옮기면서 구분자가 변경되어도 domain에 해당하는 Cars는 변경이 필요없을 것 같습니다!

jjj0611 replied:
ditto

tco0427 replied:
일급컬렉션의 장점 중 하나의 불변 컬렉션 보장이라고 학습하였습니다!! 따라서 생성자를 통한 initCar() 호출 이외의 변경에 영향을 주는 메소드를 두지 않았으므로 List<Car>에 대한 별도의 final 선언은 불필요하다고 생각합니다!!

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,67 @@
+
+public class Cars {
+
+    private static final String SEPARATOR = ",";
```
자 이제 고민을 한 번만 더 해보면 좋을 것 같아요!
지금은 carNames라는 String 덩어리를 받아서 Cars의 생성시점에 직접 split을 하여 생성을 하고 있는데,
그렇기 때문에 Cars에 ","라는 view와 관련된 매직넘버가 존재하게 되는 것 같아요.
String carNames -> List carNames로 변환하는 로직이 view에 들어가면 어떤가요?
중복 체크 등의 로직은 Cars의 역할이지만, String -> List의 변환까지 Cars의 도메인인가 하는 고민을 한 번 해보면 좋을 것 같아요 :)

jjj0611 replied:
자동차 이름의 입력 구분값이 ,에서 /로 바뀌면 domain이 변경되어야 하는가 view가 변경되어야 하는가?를 고민해보시면 좋지 않을까요?

tco0427 replied:
좋은 말씀이라고 생각합니다 :) 적용해보았습니다!!ㅎㅎ

jjj0611 commented on src/main/java/racingcar/domain/GameTotalCount.java:
```diff
@@ -0,0 +1,51 @@
+
+    private int totalAttemptCount;
+
+    private GameTotalCount(int totalAttemptCount) {
+        validatePositiveNumber(totalAttemptCount);
+        this.totalAttemptCount = totalAttemptCount;
+    }
+
+    public static GameTotalCount createGameTotalCount(String attempt) {
+        int number = translateInteger(attempt);
+
+        return new GameTotalCount(number);
+    }
```
잘 반영해주셨네요! 😀

tco0427 replied:
감사합니다!!

jjj0611 commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,66 @@
+
+    private Cars cars;
+    private int size;
+    private int winnerPosition;
```
우승자의 위치를 이렇게 따로 놓을 필요가 있을까요?
각 테스트에서 가장 많이 진행시킨 사람의 위치를 우승자의 위치라고 생각할 수 있지 않나요??

tco0427 replied:
맞습니다! 그런 의미에서 현재 존재하는 테스트에서는 공통되게 모두 0번 인덱스가 가장 많이 진행된 사람의 위치이고 따라서 이를 상수로 치환했는데요..향후에 테스트 코드가 추가됨에 따라 우승자가 0번 인덱스가 아닐 수도 있기 때문에 오해의 소지가 있는 상수라고 생각됩니다..!!
그래서 숫자 '0'을 그대로 사용하는 것도 나쁘지 않다고 생각이 들어 이렇게 수정해보았습니다!!

jjj0611 reviewed (changes requested):
안녕하세요 디우! 리뷰어 제이입니다!

말씀드린 내용 잘 반영해주셨네요! 거의 다 왔습니다!
조금 더 고민해보면 좋을 부분 코멘트 드렸으니 확인해주시고 반영해주세요 :)

tco0427 committed:
refactor: 게임 진행 여부 와 시도 횟수 감소 분리

게임 진행 여부와 시도 횟수를 감소시키는 메소드를 완전히 분리하였습니다.

tco0427 committed:
refactor: domain과 view 책임 분리

자동차 이름들을 domain에 해당하는 Cars가 받아서 이를 분리하는 것이 아니라 view에서 이미 String[] 형태로 넘겨주는 것이 적절하다. 향후에 구분자가 '/' 로 변경된다고 해서 domain에 해당하는 Cars의 코드가 변경될 이유는 없다.

tco0427 committed:
refactor: 차 생성 로직(initCar)의 순서 변경

사소한 차이이지만 검증을 먼저 하고 나서 ArrayList를 생성하는게 가독성 측면에서도 그리고 작지만 메모리 관리 측면에서도 좋다고 생각되어 순서를 변경하였습니다.

tco0427 committed:
refactor: 상수 제거

향후 추가되는 테스트에 따라서 우승자의 위치가 0번 인덱스가 아닐 수도 있으므로 각 테스트에서 가장 많이 진행시킨 사람의 위치를 우승자의 위치로 사용하도록 상수 제거

tco0427 committed:
refactor: View에 CarDto를 넘겨주도록 리팩터링

향후에 view에서 표시할 데이터가 늘어나면 기존코드에서는 printProgress()에 넘길 인자가 계속해서 늘어나게 될 것이다. 따라서 CarDto를 만들고 view에서 필요한 데이터를 가지는 CarDto를 넘기도록 리팩터링하였습니다.

tco0427 committed:
refactor: CarName에 final 키워드 추가

변경 가능성 없는 CarName 의 name 필드에 final 추가

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,65 @@
+        return cars.get(index);
+    }
+
+    private void initCar(String[] carNames) throws IllegalArgumentException {
```
처음부터 배열이 아닌 List로 넘겨주면 어떨까요?

tco0427 replied:
넵! `배열 대신 java collection을 사용하라` 라는 피드백을 프리코스 때도 받았었는데요...다양한 api를 활용할 수 있어서 좋다고 이해하긴 했는데..이런 api가 불필요한 경우에도 통일성을 위해서 모두 collection을 쓰는게 좋은지..아직은 알쏭달쏭한 느낌이라..그냥 배열을 사용했는데요,,다음 미션에서는 페어와 논의하고 답을 찾아보도록 하겠습니다!! 감사합니다.😄

jjj0611 commented on src/main/java/racingcar/domain/GameTotalCount.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public boolean isContinue() {
+        if (totalAttemptCount == GAME_END_CONDITION) {
+            return false;
+        }
+
+        return true;
```
return totalAttemptCount != GAEM_END_CONDITION
과 같이 간단하게 표현할 수 있지 않을까요?

tco0427 replied:
해당 내용도 고민해봤는데..로직이 바로 읽히는게 저는 위의 코드였어서 사용했는데요..!! 한 번 고민해보겠습니다!!

jjj0611 commented on src/main/java/racingcar/dto/CarDto.java:
```diff
@@ -0,0 +1,26 @@
+
+public class CarDto {
+
+    private String carName;
+    private int position;
```
final이 될 수 있지 않을까요? :)

tco0427 replied:
동의합니다!!

jjj0611 commented on src/main/java/racingcar/dto/CarDto.java:
```diff
@@ -0,0 +1,26 @@
+
+import racingcar.domain.Car;
+
+public class CarDto {
```
Dto까지 잘 분리를 해주셨네요 :)

jjj0611 reviewed (approved):
안녕하세요 디우 리뷰어 제이입니다.
자동차 경주게임 코드리뷰를 진행하면서 코드가 많이 개선된 것들이 눈에 보이네요.
학습해야할 것들에 대해서 잘 학습하고, 리뷰 코멘트도 잘 반영해주셔서 감사해요 :)

코멘트 조금 남겨드렸으니 마지막으로 확인해주세요!
그동안 리뷰 드린 것들이 많은 도움이 되었으면 좋겠네요 :)
자동차 경주게임 고생하셨고, PR은 여기서 머지하겠습니다.
다음 미션도 화이팅하세요! 😄

tco0427 commented:
짧다면 짧고 길다면 긴 일주일이 조금 넘는 기간동안 꼼꼼히 리뷰해주신 느낌이 들어서 정말 감사했습니다!!
또 많은 부분 배워가는거 같습니다😄 (더 많이 배우고 싶었는데..아쉽네요ㅠ.ㅠ)
제가 아직은 부족해서..제이께서 피드백 주신 내용을 200% 흡수하지 못한 것 같은데요..이해를 돕기 위해서 늦은 시간임에도 불구하고 통화까지 해주셔서 피드백 주시는 진심에 감동했습니다...감사합니다!!
